compiler/transforms/schedule_memref_linalg.py

```diff
@@ -0,0 +1,302 @@
+from xdsl.context import MLContext
+from xdsl.dialects import builtin, memref_stream, stream
+from xdsl.ir import Block, Region
+from xdsl.ir.affine import AffineConstantExpr, AffineDimExpr, AffineExpr, AffineMap
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.rewriter import InsertPoint
+
+from compiler.util.canonicalize_affine import canonicalize_map
+
+
+def disable_dims(map: AffineMap, dim: int) -> AffineMap:
+    """
+    Returns an affine map with the first `dim` dimensions set to 0
+
+    For example:
+        (d0, d1, d2) -> d0 + d1 + d2
+    For `dim` = 1, will return:
+        (d1, d2) -> d1 + d2
+    For `dim` = 2, will return:
+        (d2) -> d2
+    """
+    return canonicalize_map(
+        map.replace_dims_and_symbols(
+            [AffineConstantExpr(0) for _ in range(dim)]
+            + [AffineDimExpr(i) for i in range(map.num_dims - dim)],
+            [],
+            map.num_dims - dim,
+            0,
+        )
+    )
+
+
+def rotate_dims(map: AffineMap, dim: int) -> AffineMap:
+    """
+    Returns an affine map with the first `dim` dimensions rotated
+
+    For example:
+        (d0, d1, d2) -> 1 * d0 + 2 * d1 + 3 * d2
+    For `dim` = 3, will return:
+        (d0, d1, d2) -> 3 * d0 + 1 * d1 + 2 * d2
+    For `dim` = 2, will return:
+        (d0, d1, d2) -> 2 * d0 + 1 * d1 + 3 * d2
+    """
+    new_dims = [AffineDimExpr(i) for i in range(dim)]
+    # rotate dims by popping first and appending
+    new_dims.append(new_dims.pop(0))
+    # keep remaining dims
+    new_dims = new_dims + [AffineDimExpr(i) for i in range(dim, map.num_dims)]
+    return canonicalize_map(
+        map.replace_dims_and_symbols(new_dims, [], len(new_dims), 0)
+    )
+
+
+def rotate_bounds(bounds: list[int | None], dim: int) -> list[int | None]:
+    """
+    Returns the bounds after rotating dims, as in rotate_dims
+    """
+    bounds = bounds.copy()
+    bounds.insert(0, bounds.pop(dim - 1))
+    return bounds
+
+
+def split_dim(map: AffineMap, dim: int, template_bound: int) -> AffineMap:
```

Comment:
Your guide refers to this as tiling, but e.g. halide calls 2D loop splitting tiling, and 1D loop splitting just splitting.
Maybe change `tiling` `->` `splitting` there?

Reply:
yes!